processor/src/main/java/com/linecorp/decaton/processor/metrics/Metrics.java

```diff
@@ -59,36 +59,30 @@ private Metrics(AvailableTags availableTags) {
      * the instance from {@link MeterRegistry} when the last reference to the meter closed and disappeared.
      */
     abstract static class AbstractMetrics implements AutoCloseable {
-        private static final Map<Meter.Id, AtomicInteger> meterRefCounts = new HashMap<>();
-        private final List<Meter> meters;
-
-        protected AbstractMetrics() {
-            meters = new ArrayList<>();
-        }
-
-        <T extends Meter> T meter(Supplier<T> ctor) {
-            synchronized (meterRefCounts) {
-                T meter = ctor.get();
-                meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger())
-                              .incrementAndGet();
-                meters.add(meter);
-                return meter;
-            }
+        private static final ConcurrentMap<Id, AtomicInteger> meterRefCounts = new ConcurrentHashMap<>();
+        private final List<Meter> meters = new ArrayList<>();
+
+        synchronized <T extends Meter> T meter(Supplier<T> ctor) {
+            T meter = ctor.get();
+            meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger())
```

Comment:
It would be (by current usage) not realistic to happen but this concurrency semantics feels bit broken because it would leaks a reference count in following situation?

A and B are operating on different `XXMetrics extends AbstractMetrics` instance.

1. Thread A ` AtomicInteger count = meterRefCounts.get(id); `
2. Thread B `meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger()) => returns existing`
3. Thread A `count.decrementAndGet() <= 0) && meterRefCounts.remove(id);`
4. Thread B ` .incrementAndGet();`

So I think it's better to keep `meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger()).incrementAndGet()` this whole operation atomic among all `XXMetrics` instances by keeping synchronize on `meterRefCounts`.

Reply:
Okay. I think we can use `ConcurrentMap.computeIfPresent` which is atomic.
I don't want to nest synchronization

Reply:
You can just get rid of method level synchronize in case? Why do we needed to replace `synchronized` on `refcount` with method level synchornize in the first place?

Reply:
Ah, I see what you mean. We can indeed synchronize only on this static `meterRefCounts`